src/tasks.ts

```diff
@@ -0,0 +1,102 @@
+import { spawnSync } from "child_process";
+import * as path from "path";
+import { Range, TextEditor } from "vscode";
+import { outputChannel, ws } from "./extension";
+import { TaskOutput } from "./taskRunner";
+import { NameToThemeNumber } from "./themePicker";
+import { util } from "./utility";
+
+/**
+ * D2Tasks - static functions to be used in tasks.  Functions need
+ * to be synchronous, and return their results in a callback.
+ */
+class D2Tasks {
+  public compile(text: string, log?: TaskOutput, err?: TaskOutput): string {
+    const layout: string = ws.get("previewLayout", "dagre");
+    const theme: string = ws.get("previewTheme", "default");
+    const sketch: boolean = ws.get("previewSketch", false);
+    const themeNumber: number = NameToThemeNumber(theme);
+    const d2Path: string = ws.get("execPath", "d2");
+
+    err?.("Starting Compile...");
+    err?.(`Layout: ${layout}  Theme: ${theme}  Sketch: ${sketch}\r\n`);
+
+    const args: string[] = [
+      `--layout=${layout}`,
+      `--theme=${themeNumber}`,
+      `--sketch=${sketch}`,
+      "-",
+    ];
+    const proc = spawnSync(d2Path, args, { input: text, encoding: "utf-8" });
+
+    /** proc.status: 0 - success
+     *  proc.status: 1 - errors
+     *  proc.pid: 0 - EXE Not Found
+     */
+    if (proc.pid === 0) {
+      util.showErrorToolsNotFound(proc.error?.message ?? "");
+    } else {
+      const error: string = proc.stderr?.toString() ?? "";
+
+      for (const msg of error.split("\n")) {
+        if (msg.length === 0) {
+          // Sometimes there is a blank line that we
+          // don't have to worry about
+          continue;
+        }
+
+        err?.(msg, true);
+      }
+    }
+
+    err?.("");
+
+    let data = "";
+    // Success! Read the output
+    if (proc.status === 0) {
+      data = proc.stdout.toString();
+    }
+
+    return data;
+  }
+
+  format(textEditor: TextEditor): void {
+    const fileText = textEditor.document.getText();
+
+    const d2Path: string = ws.get("execPath", "d2");
+    const proc = spawnSync(d2Path, ["fmt", "-"], { input: fileText });
+
+    let errorString = "";
+    if (proc.status !== 0) {
+      errorString = proc.stderr.toString();
+      outputChannel.appendError(errorString);
+      return;
+    }
+
+    const data: string = proc.stdout.toString();
```

Comment:
I was thinking this should just be up to this point to get the formatted string (similarly to compile above), and then the updating of the text editor would be handled outside of this (similarly to how the updating of the preview is outside of compile)

Reply:
I pass the TextEditor into the format function, so it makes more sense to do all operations on it there.  If I pull that out and only pass the string, the function almost becomes empty.

Reply:
ok I think it is good for now and we can always refactor if necessary